docs/contributing.rst

```diff
@@ -199,3 +199,121 @@ The documentation can be built by running::
+contributors.
+
+Merging pull requests
+~~~~~~~~~~~~~~~~~~~~~
+
+If at all possible, pull requests submitted by an external contributor should be
+reviewed and approved by all core developers before being merged. Pull requests
+submitted by a core developer should be reviewed and approved by all other core
+developers before being merged.
+
+Pull requests should not be merged until all CI checks have passed (Travis, AppVeyor,
+Coveralls) against code that has had the latest master merged in.
+
+Compatibility and versioning policies
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Because Zarr is a data storage library, there are two types of compatibility to
+consider: API compatibility and data format compatibility.
+
+Here we consider all functions, classes and methods that do not begin with an
+underscore as part of the Zarr public API. Any change to the public API that does **not**
+break existing third party code importing Zarr, or cause third party code to behave in
+a different way, is a **backwards-compatible API change**. For example, adding a new
+function, class or method is usually a backwards compatible change. However, removing a
+function, class or method; removing an argument to a function or method; adding a
+required argument to a function or method; or changing the behaviour of a function or
+method, are examples of **backwards-incompatible API changes**.
+
+If a release contains no changes to the public API (e.g., contains only bug fixes or
+other maintenance work), then the micro version number should be incremented (e.g.,
+2.2.0 -> 2.2.1). If a release contains API changes, but all API changes are
+backwards-compatible, then the minor version number should be incremented
+(e.g., 2.2.1 -> 2.3.0). If a release contains any backwards-incompatible API changes,
+the major version number should be incremented (e.g., 2.3.0 -> 3.0.0).
+
+Exceptions can be made for any function, class or method which has been documented as
+an experimental feature, i.e., backwards-incompatible changes can be included in a
+minor release, although this should be avoided wherever possible.
+
+The data format used by Zarr is defined by a specification document, which should be
+platform-independent and contain sufficient detail to construct an interoperable
+software library to read and/or write Zarr data using any programming language. The
+latest version of the specification document is available from the :ref:`spec` page.
+Here, **data format compatibility** means that all software libraries that implement a
+particular version of the Zarr storage specification are interoperable, in the sense
+that data written by any one library could be read by all others. It is obviously
+desirable to maintain data format compatibility wherever possible. However, if a change
+is needed to the storage specification, and that change would break data format
+compatibility in any way, then the storage specification version number should be
+incremented (e.g., 2 -> 3).
```

Comment:
What happens when we add something to the data format spec which is not breaking? Is there a way to version this?

Also related how do we migrate between these enhancement versions? Lastly do we have a way to check the data format spec version of Zarr and/or the files themselves?

Reply:
At the moment, I cannot imagine any change to the data format spec that does not break data format compatibility in some way, in the sense that data written by any one implementation can be read correctly by all others. Hence the storage spec has only a major version number, there is no minor or micro version number. This simplifies the definition of data format compatibility, i.e., all libraries implementing a given spec version number should be fully interoperable. That said, if you have an example of a spec change that does not fit into this model, happy to discuss.

The spec version of the data files is included in the metadata (.zarray or .zgroup). The spec version that zarr implements is currently stored as the ``ZARR_FORMAT`` variable in the ``zarr.meta`` module. There is a check on reading array or group metadata, if there is a difference between the spec version in the metadata files and the ``ZARR_FORMAT`` variable then an exception is raised.

When we moved from spec version 1 to 2 I implemented a function ``zarr.storage.migrate_1to2`` which updates the metadata resources to conform to spec version 2. I imagined that we would do something similar for any future spec changes.

Reply:
As an example, we recently opted to not always write metadata files and noted this in the spec. We have also made sure that current Zarr treats this as empty metadata. That said, I'm not sure if an older version of Zarr (claiming to support the same spec version) would handle this correctly. Even though what we added didn't fundamentally break the version 2 spec. Instead it provided a superset of the existing functionality. IOW this seems like a feature we added to the spec that would make it 2.1. Though this could just be me being pedantic. :)

Everything else sounds great. We might want to mention the `ZARR_FORMAT` variable here.

Reply:
It's a fair point, although that change was a bit of a grey area in that [zarr 2.1 is happy if .zattrs is missing](https://github.com/zarr-developers/zarr/blob/v2.1.4/zarr/attrs.py#L71) and [the spec did not say whether .zattrs is required or not](https://github.com/zarr-developers/zarr/blob/v2.1.4/docs/spec/v2.rst#attributes), so saying that .zattrs is optional was a clarification and did not break compatibility with previous zarr versions implementing spec version 2. I might have been a bit more cautious if I knew there were other zarr implementations out there, but it still would have been a clarification rather than a change to the spec.

We may still come across other cases which are more clear cut, in that some change is required in the spec to support a new feature, and where using something more fine-grained than a single major version number for the spec version could be useful, but I find it hard to think that through without a concrete case in front of me. The questions we would need to answer are things like, what kinds of changes would be allowed in a minor spec change (e.g., 2.0 -> 2.1), how should an implementation behave if it implements spec version 2.0 and encounters data using spec version 2.1, etc. I think we can cross those bridges if and when the need arises.

Reply:
We could probably smooth this over by adding one last patch release to 2.1 with this change. Alternatively we could make the behavior something users opt-in to. That said, I'm not too worried about it. In any event, this discussion is diverging from the spec version and can be continued in an issue if we would like.

Sure, make sense. Easy to add once we need it.